### ISSUE TITLE
What's the best way to report loss metrics on test set in Tensorboard?

### ISSUE THREAD

**Issue by thomasjanssens** [facebookresearch/detectron2#499]
I'm training Mask R-CNN on a custom dataset using _plain_train_net.py_ as a base.
What would be the best way to report validation loss metrics in Tensorboard? Should I write a custom evaluator, just run through the validation set without updating the parameters in the training loop, or something else? Ideally, I'd like to get all validation  COCO and loss metrics in Tensorboard, which I suppose means I need to put them all in EventStorage.

Any pointers or examples I could look at?

**Comment by ppwwyyxx** [comment]
There is no definition of test-time loss. If you have made such a definition you can implement it in a custom model.

If you just want the training loss on the test data, you can just replace the training data with test set data and the model output are the losses.

**Comment by ortegatron** [comment]
Hi, I have written a custom trainer plotting the loss on validation, please have a look here, hope it helps you:
https://medium.com/@apofeniaco/training-on-detectron2-with-a-validation-set-and-plot-loss-on-it-to-avoid-overfitting-6449418fbf4e

**Comment by vishal00100** [comment]
@ortegatron I might be missing something but don't you need to set model to *eval* mode first?
(See [SO](https://stackoverflow.com/questions/55627780/evaluating-pytorch-models-with-torch-no-grad-vs-model-eval)).

`with torch.no_grad` is recommended by not necessary.
